### PR TITLE
Fix some DMA SPI issues with DMA IP v3 and/or word size > 8 bits

### PR DESCRIPTION
**Pull request by multiplemonomials**
### Summary of changes 

Working on STM32H5 SPI support, I dug a bit deeper into some DMA SPI errors that we were seeing, and ended up finding and fixing a few bugs.  These were:
- in stm_dma_utils.c, some of the dest and src constants were swapped.  This had the effect of making app DMA IP v3 transfers increment the source address if you set the destination increment flag.  Oops!  Frankly I have no idea how it ever passed any tests at all like this, it should never have worked
- DMA_DEST_DATAWIDTH_WORD replaced by DMA_DEST_DATAWIDTH_BYTE in 2 places (spotted by @wdx04 in #196).  This could mess up the DMA system if you requested a 32-bit transfer
- If the word size is set to 16 or 32 bits, it turns out that the STM32 DMA HAL requires the DMA's transfer size to be set accordingly.  We were not doing this, causing any attempt to do wider transfers using DMA (on some STM processors at least) to fail.
- Initializing an SPI slave with MISO set to NC would print a warning, even though it did work as expected.  This warning has been removed.

### Documentation 

None

----------------------------------------------------------------------------------------------------------------
### Pull request type 

    [X] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results 

    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [X] Tests / results supplied as part of this PR
    
 On STM32H5, spi-basic test now passes completely!  Need to run tests on other devices.
----------------------------------------------------------------------------------------------------------------

### CODE REVIEW

**Comment by chrissnow** [comment]
@multiplemonomials I think you may have missed a file in your commit.
```
[build] Starting build
[proc] Executing command: "C:\Program Files\CMake\bin\cmake.EXE" --build c:/Code/test/build/VIPER_EV-Debug --config Debug --target all --
[build] [22/175   0% :: 1.190] Building C object mbed-os/CMakeFiles/mbed-os.dir/targets/TARGET_STM/stm_spi_api.c.obj
[build] FAILED: mbed-os/CMakeFiles/mbed-os.dir/targets/TARGET_STM/stm_spi_api.c.obj 
[build] "C:\Program Files\ccache-4.10.1-windows-x86_64\ccache.exe" C:\PROGRA~2\ARMGNU~1\125863~1.3RE\bin\AR19DD~1.EXE @mbed-os\CMakeFiles\mbed-os.dir\targets\TARGET_STM\stm_spi_api.c.obj.rsp -MD -MT mbed-os/CMakeFiles/mbed-os.dir/targets/TARGET_STM/stm_spi_api.c.obj -MF mbed-os\CMakeFiles\mbed-os.dir\targets\TARGET_STM\stm_spi_api.c.obj.d -o mbed-os/CMakeFiles/mbed-os.dir/targets/TARGET_STM/stm_spi_api.c.obj -c C:/Code/test/mbed-os/targets/TARGET_STM/stm_spi_api.c
[build] C:/Code/test/mbed-os/targets/TARGET_STM/stm_spi_api.c: In function 'configure_dma_transfer_size':
[build] C:/Code/test/mbed-os/targets/TARGET_STM/stm_spi_api.c:1565:41: error: 'dmaHandle' undeclared (first use in this function)
[build]  1565 |     uint32_t * const transferSizePtr = &dmaHandle->Init.PeriphDataAlignment;
[build]       |                                         ^~~~~~~~~
[build] C:/Code/test/mbed-os/targets/TARGET_STM/stm_spi_api.c:1565:41: note: each undeclared identifier is reported only once for each function it appears in
[build] C:/Code/test/mbed-os/targets/TARGET_STM/stm_spi_api.c:1588:9: error: 'transferSizePtr1' undeclared (first use in this function); did you mean 'transferSizePtr2'?
[build]  1588 |     if(*transferSizePtr1 != neededSizeVal1 || *transferSizePtr2 != neededSizeVal2)
[build]       |         ^~~~~~~~~~~~~~~~
[build]       |         transferSizePtr2
[build] C:/Code/test/mbed-os/targets/TARGET_STM/stm_spi_api.c:1565:22: warning: unused variable 'transferSizePtr' [-Wunused-variable]
[build]  1565 |     uint32_t * const transferSizePtr = &dmaHandle->Init.PeriphDataAlignment;
[build]       |                      ^~~~~~~~~~~~~~~
```

File: targets/TARGET_STM/stm_spi_api.c [hunk 1]
```diff
@@ -1536,6 +1535,63 @@ typedef enum {
...
+        neededSizeVal1 = DMA_DEST_DATAWIDTH_HALFWORD;
+        neededSizeVal2 = DMA_SRC_DATAWIDTH_HALFWORD;
+    }
+    else
+    {
+        neededSizeVal1 = DMA_DEST_DATAWIDTH_WORD;
+        neededSizeVal2 = DMA_SRC_DATAWIDTH_WORD;
+    }
+#else
+    uint32_t * const transferSizePtr = &dmaHandle->Init.PeriphDataAlignment;
```
Comment (chrissnow) [inline targets/TARGET_STM/stm_spi_api.c:1565]: transferSizePtr should be transferSizePtr1
dmaHandle should be dmaChannel

**Comment by chrissnow** [comment]
With the  above changes I can confirm 16-bit works on STM32WB55.

**Review by chrissnow — CHANGES_REQUESTED** [review]
Does not compile if the DMA controller is !=DMA_IP_VERSION_V3

With the above changes I can confirm 16-bit works on STM32WB55.

File: targets/TARGET_STM/stm_spi_api.c (hunk 1 above)
Comment (multiplemonomials) [reply targets/TARGET_STM/stm_spi_api.c]: Good catch, thank you!

**Review by chrissnow — APPROVED** [review]

**Review by JohnK1987 — APPROVED** [review]

**Pull request merged**